File: thanh123456_AcceptedCodeVNSPOJ/CREC01.cpp
#include <bits/stdc++.h>
#define pb push_back
#define p push
#define mp make_pair
using namespace std;
const int N = 1e3 + 100;
int n,m;
char c[N][N];
int l[N],r[N],d[N];
void nhap()
{
    scanf("%d %d\n",&n,&m);
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= m; j++)
        {
            scanf("%c",&c[i][j]);
        }
        scanf("\n");
    }
}
void solve()
{
    long long res = 0;
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= m; j++)
        {
            if (c[i][j] == '1') {d[j]++;}
            else {d[j] = 0;}
        }
        l[0] = 0;
        for (int j = 1; j <= m; j++)
        {
            l[j] = j;
            int k = j - 1;
            while (k > 0 && d[j] < d[k]) {
                l[j] = l[k];
                k = l[k] - 1;
            }
        }
        for (int j = m; j >= 1; j--)
        {
            r[j] = j;
            int k = j + 1;
            while (k <= m && d[j] <= d[k])
            {
                r[j] = r[k];
                k = r[k] + 1;
            }
        }
        for (int j = 1; j <= m; j++)
        {
            res+=1ll * (j - l[j] + 1) * (r[j] - j + 1) * d[j];
        }
    }
    cout<<res;
}
int main()
{
    nhap();
    solve();
}

File: thanh123456_AcceptedCodeVNSPOJ/CREC01.pas
const
     fin='';fon='';
     maxn=2000;
type
     mang1=array [0..maxn,0..maxn] of char;
     mang2=array [0..maxn] of longint;
var
     n,m,top:longint;
     a:mang1;
     d,l,r,stack:mang2;
procedure nhap;
     var
          i,j:longint;
     begin
          assign(input,fin);reset(input);
          assign(output,fon);rewrite(output);
          readln(n,m);
          for i:=1 to n do
               begin
                    for j:=1 to m do read(a[i,j]);
                    readln;
               end;
     end;
procedure push(x:longint);
     begin
          inc(top);
          stack[top]:=x;
     end;
Function Get:longint;
     begin
          exit(stack[top]);
     end;
Procedure pop;
     begin
          stack[top]:=0;
          dec(top);
     end;
procedure solve;
     var
          i,j,res:longint;
          s:int64;
     begin
          s:=0;
          for i:=1 to n do
               begin
                    for j:=1 to m do
                         begin
                              if a[i,j]='1' then inc(d[j])
                              else d[j]:=0;
                         end;
                    top:=0;
                    push(1);
                    l[1]:=1;
                    for j:=2 to m do
                         begin
                              res:=0;
                              while (top>0) and (d[j]<d[get]) do
                                   begin
                                        res:=res+l[get];
                                        pop;
                                   end;
                              l[j]:=res+1;
                              push(j);
                         end;
                    top:=0;
                    push(m);
                    r[m]:=1;
                    for j:=m-1 downto 1 do
                         begin
                              res:=0;
                              while (top>0) and (d[j]<=d[get]) do
                                   begin
                                        res:=res+r[get];
                                        pop;
                                   end;
                              r[j]:=res+1;
                              push(j);
                         end;
                    for j:=1 to m do
                         begin
                              //writeln(l[j],' ',r[j]);
                              if a[i,j]='0' then continue;
                              s:=s+int64(l[j]*r[j]*d[j]);
                         end;
                    //writeln(s);
               end;
          writeln(s);
     close(input); close(output);
     end;
begin
     nhap;
     solve;
end.
